controllers/compaction/reconciler.go

```diff
@@ -107,12 +107,39 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	logger := r.logger.WithValues("etcd", kutil.Key(etcd.Namespace, etcd.Name).String())
 
+	// Update metrics for currently running compaction job, if any
+	job := &batchv1.Job{}
```

Comment:
Why are we fetching the job here and checking the status (active, succeeded, failed) when we're already doing it in `reconcileJob()`? This seems redundant.

The PR may be correcting the inaccuracy in metrics exported, but in the process introduces redundancies and inefficiency in the reconciliation loop. I would ask you to please take a more holistic approach for this PR - if done right, it can solve both #685 as well as #698, by correcting the reconciliation loop flow (`reconcileJob()` should be called even if the revision difference crosses the threshold, unlike how it is done today). Ideally, reconciliation loop logic should be sound, and only then metrics should be exposed based on the operations and status information from the job.

Reply:
We are checking the job status here because `reconcileJob()` is only called when the revision difference crosses the threshold. If the Job status is active, it also don't check the revision difference unnecessarily.

I don't understand what do you mean by `reconcileJob() should be called even if the revision difference crosses the threshold, unlike how it is done today` ! Currently, reconcileJob() is called upon revision difference crossing the threshold ..

What do you mean by holistic approach? You can point out which case is not handled in https://github.com/gardener/etcd-druid/issues/685 and https://github.com/gardener/etcd-druid/issues/698. I will add the case in this PR, if not added already.

Reply:
> We are checking the job status here because reconcileJob() is only called when the revision difference crosses the threshold. If the Job status is active, it also don't check the revision difference unnecessarily

This is exactly what needs to be changed. reconcileJob() needs to run irrespective of the difference in revisions. Meaning, the revision difference needs to be checked within reconcileJob(), which then determines whether to deploy a new job or not. But the check for the currently running job needs to happen irrespective of the revision check.

```code
reconcile() {
  ...
  reconcileJob()
}

reconcileJob() {
  getJob()
  if job exists {
    checkRunningJobStatus()
    updateJobStatusMetrics()
    if jobStatus == Running {
      return
    }
    if jobStatus == Succeeded || jobStatus == Failed {
      deleteJob()
    }
  }

  if deltaSnapRev - fullSnapRev > threshold {
    deployJob()
  }
}
```

This ensures that the reconciliation loop logic is sound, and also takes care of metrics being correctly exported.

Reply:
Okay. Now I understand why do you think it's redundancy !

See, deployJob() can also fail instantly (for some reason). That's why we need to check job status even after deployJob() so that the controller is not run again from start. Only redundant check for  `job` is in line number 195. I think that is required after fetching leases and checking diff. One extra job status check won't be too much.

Otherwise , it is just moving codes from one block to other.

Reply:
> deployJob() can also fail instantly (for some reason)

For what reason exactly? And why would that have any effect on checking the job status?

> That's why we need to check job status even after deployJob() so that the controller is not run again from start

Once the reconciler deploys the job, it must requeue and exit, so that in the next reconciliation it can again check the job status and take appropriate actions. 

> One extra job status check won't be too much

It's not about whether it's too much. Even though the network impact of this may me minimal, the fact that the reconciler checks the same job status twice seems unnecessary to me, when in fact we can simplify the reconciliation flow logic to ensure that the job status is fetched irrespective of the revision threshold breach, and deploy a new job only if the threshold is breached and a job is not already currently running. I think this is a valid ask.

Reply:
I tested repeatedly and after facing various error conditions under various circumstances , I decided this flow. I can't list them now as I don't remember all of them. What you suggesting is movement of code block from one place to another. I had my rationale and I tried to save a few loop reconciliation. But one or two redundant check of job status won't do any harm, specially when the job status comes from informer cache. Current flow goes like this:

1. In the Reconciler, a job is first checked for status. If it's succeeded, a delete call is issued on job.
2. The flow then enters to fetching leases and check the diff between leases.  Ideally, for most cases, when the reconciler is called upon status change of job for succeeded, diff check would not cross threshold because the delta events are just compacted. But, there might be some busy cluster where a job status change to successful and diff threshold cross may happen simultaneously.
3. Then we need to check if the job is really deleted or not and so we need the "redundant" job check of line number 195.
4. After the `createCompactionJob` call, I am no more issuing `Get` call on job, other than while deleting.
5. You may think line number 248 is redundant but I placed that for a reason. This line will only execute when a job is instantaneously succeeded, may be if there is no full snapshot in the storage at all or may be for some unknown reason which we should always be prepared for in a distributed system.
6. Action upon `job.Status.Failed` is executing only once, it's not redundant.

As I said, I won't feel confident if I need to change the flow that I wrote because I came to it after lots of testing in real scenarios upon various favourable and erroneous conditions. But if you just wish to move the codes from line number 110 inside the function `reconcileJob()`, that I can do. Do you wish that?

Reply:
> What you suggesting is movement of code block from one place to another

Not really. I suggested not adding the code block in this PR as it already exists in the code, within `reconcileJob()` method. I simply suggested calling `reconcileJob()` irrespective of the revision threshold breach, because reconciliation of a job is not just about deploying the job, but also fetching the status of the job and deleting the job, and these tasks need to performed irrespective of whether the revision threshold has been breached or not. Only the task of deploying a new job should be dependent on the revision threshold breach.

regarding the flow, I would request a second opinion from @unmarshall , because to me the flow you mentioned still seems to be repeated certain code like getJob() and checkStatus() partially and deleteJob(), and seems to be trying to handle specific corner cases, rather than being generic enough to work for any case.

/needs second-opinion
/assign @unmarshall

Reply:
I made the changes